File: java-bases/base/src/main/java/cn/wjybxx/base/mutable/MutableDouble.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.base.mutable;

/**
 * @author wjybxx
 * date - 2024/1/4
 */
public class MutableDouble extends Number implements MutableNumber<Double>, Comparable<MutableDouble> {

    private double value;

    public MutableDouble() {
    }

    public MutableDouble(double value) {
        this.value = value;
    }

    public MutableDouble(Number value) {
        this.value = value.doubleValue();
    }

    public MutableDouble(String value) {
        this.value = Double.parseDouble(value);
    }

    // region cast

    @Override
    public Double getValue() {
        return value;
    }

    @Override
    public void setValue(Double value) {
        this.value = value;
    }

    public void setValue(final double value) {
        this.value = value;
    }

    @Override
    public int intValue() {
        return (int) value;
    }

    @Override
    public long longValue() {
        return (long) value;
    }

    @Override
    public float floatValue() {
        return (float) value;
    }

    @Override
    public double doubleValue() {
        return value;
    }

    // endregion


    // region op

    /** 加上操作数 */
    public void add(final double operand) {
        this.value += operand;
    }

    /** 返回加上操作数后的值 */
    public double addAndGet(final double operand) {
        this.value += operand;
        return value;
    }

    /** 返回当前值 */
    public double getAndAdd(final double operand) {
        final double last = value;
        this.value += operand;
        return last;
    }

    /** 加1 */
    public void increment() {
        value++;
    }

    /** 返回加1后的值 */
    public double incrementAndGet() {
        return ++value;
    }

    /** 加1并返回当前值 */
    public double getAndIncrement() {
        return value++;
    }

    /** 减1 */
    public void decrement() {
        value--;
    }

    /** 返回减1后的值 */
    public double decrementAndGet() {
        return --value;
    }

    /** 减1并返回当前值 */
    public double getAndDecrement() {
        return value--;
    }

    @Override
    public MutableDouble add(Number operand) {
        this.value += operand.doubleValue();
        return this;
    }

    @Override
    public MutableDouble subtract(Number operand) {
        this.value -= operand.doubleValue();
        return this;
    }

    @Override
    public MutableDouble multiply(Number operand) {
        this.value *= operand.doubleValue();
        return this;
    }

    @Override
    public MutableDouble divide(Number operand) {
        this.value /= operand.doubleValue();
        return this;
    }

    // endregion

    // region equals

    @Override
    public int compareTo(MutableDouble that) {
        return Double.compare(value, that.value);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        MutableDouble that = (MutableDouble) o;
        return value == that.value;
    }

    @Override
    public int hashCode() {
        return Double.hashCode(value);
    }

    @Override
    public String toString() {
        return "MutableDouble{" +
                "value=" + value +
                '}';
    }

    // endregion
}

File: csharp/Wjybxx.Commons.Core/src/Mutable/MutableDouble.cs
#region LICENSE

// Copyright 2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;

#pragma warning disable CS1591
namespace Wjybxx.Commons.Mutable;

/// <summary>
/// 可变Int32
/// </summary>
public class MutableDouble : IMutableNumber<double>, IEquatable<MutableDouble>, IComparable<MutableDouble>
{
    private double _value;

    public MutableDouble(double value = 0d) {
        _value = value;
    }

    public double Value {
        get => _value;
        set => _value = value;
    }

    public int IntValue => (int)_value;
    public long LongValue => (long)_value;
    public float FloatValue => (float)_value;
    public double DoubleValue => _value;

    #region op

    /** 加上操作数 */
    public void Add(double operand) {
        this._value += operand;
    }

    /** 返回加上操作数后的值 */
    public double AddAndGet(double operand) {
        this._value += operand;
        return _value;
    }

    /** 返回当前值 */
    public double GetAndAdd(double operand) {
        double last = _value;
        this._value += operand;
        return last;
    }

    /** 加1 */
    public void Increment() {
        _value++;
    }

    /** 返回加1后的值 */
    public double IncrementAndGet() {
        return ++_value;
    }

    /** 加1并返回当前值 */
    public double GetAndIncrement() {
        return _value++;
    }

    /** 减1 */
    public void Decrement() {
        _value--;
    }

    /** 返回减1后的值 */
    public double DecrementAndGet() {
        return --_value;
    }

    /** 减1并返回当前值 */
    public double GetAndDecrement() {
        return _value--;
    }

    #endregion

    #region compare

    public int CompareTo(MutableDouble? other) {
        if (ReferenceEquals(this, other)) return 0;
        if (ReferenceEquals(null, other)) return 1;
        return _value.CompareTo(other._value);
    }

    public static bool operator <(MutableDouble? left, MutableDouble? right) {
        return Comparer<MutableDouble>.Default.Compare(left, right) < 0;
    }

    public static bool operator >(MutableDouble? left, MutableDouble? right) {
        return Comparer<MutableDouble>.Default.Compare(left, right) > 0;
    }

    public static bool operator <=(MutableDouble? left, MutableDouble? right) {
        return Comparer<MutableDouble>.Default.Compare(left, right) <= 0;
    }

    public static bool operator >=(MutableDouble? left, MutableDouble? right) {
        return Comparer<MutableDouble>.Default.Compare(left, right) >= 0;
    }

    #endregion

    #region equals

    public bool Equals(MutableDouble? other) {
        if (other == null) {
            return false;
        }
        return _value.Equals(other._value);
    }

    public override bool Equals(object? obj) {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((MutableDouble)obj);
    }

    public override int GetHashCode() {
        return _value.GetHashCode();
    }

    public static bool operator ==(MutableDouble? left, MutableDouble? right) {
        return Equals(left, right);
    }

    public static bool operator !=(MutableDouble? left, MutableDouble? right) {
        return !Equals(left, right);
    }

    public override string ToString() {
        return $"{nameof(Value)}: {Value}";
    }

    #endregion
}